File: js/packages/mechanisms/casper/test/exact/facilitator/scheme.test.ts
import casperSdk from "casper-js-sdk";
import type { TransactionV1 } from "casper-js-sdk";

const { KeyAlgorithm, PrivateKey } = casperSdk;
import { describe, expect, it } from "vitest";
import { ExactCasperScheme } from "../../../src/exact/facilitator/scheme";
import {
  ErrAmountMismatch,
  ErrExpired,
  ErrInvalidAsset,
  ErrInvalidPayTo,
  ErrInvalidScheme,
  ErrInvalidSignature,
  ErrNetworkMismatch,
  ErrNotYetValid,
  ErrPayToMismatch,
  ErrPublicKeyMismatch,
  ErrSettleFailed,
} from "../../../src/exact/facilitator/scheme";
import { ExactCasperScheme as ClientExactCasperScheme } from "../../../src/exact/client/scheme";
import { FacilitatorCasperSigner } from "../../../src/signer";
import { toClientCasperSigner } from "../../../src/signer";
import { ExactCasperPayload } from "../../../src/types";

const testAsset = "aabbccddeeff0011223344556677889900aabbccddeeff001122334455667788";
const testPayTo = "00aabbccddeeff0011223344556677889900aabbccddeeff001122334455667788";
const testNetwork = "casper:casper-test";

function createMockSigner(overrides?: {
  signTransaction?: () => Promise<void>;
  putTransaction?: () => Promise<string>;
  waitForTransaction?: () => Promise<void>;
}): FacilitatorCasperSigner {
  const privateKey = PrivateKey.generate(KeyAlgorithm.ED25519);

  return {
    getNetworkConfig: async () => ({
      chainName: "casper-test",
      rpcUrl: "http://localhost:11101/rpc",
    }),
    getAddresses: () => [privateKey.publicKey.accountHash().toHex()],
    getPublicKeyHex: () => privateKey.publicKey.toHex(),
    signTransaction: overrides?.signTransaction ?? (async () => {}),
    putTransaction: overrides?.putTransaction ?? (async () => "transaction-hash"),
    waitForTransaction: overrides?.waitForTransaction ?? (async () => {}),
  };
}

async function createValidPayload(): Promise<{
  payload: ExactCasperPayload;
  payerAddress: string;
}> {
  const privateKey = PrivateKey.generate(KeyAlgorithm.ED25519);
  const clientSigner = toClientCasperSigner(privateKey);
  const clientScheme = new ClientExactCasperScheme(clientSigner);

  const requirements = {
    scheme: "exact" as const,
    network: testNetwork,
    asset: testAsset,
    amount: "1000000",
    payTo: testPayTo,
    maxTimeoutSeconds: 300,
    extra: {
      name: "TestToken",
      version: "1",
    },
  };

  const result = await clientScheme.createPaymentPayload(2, requirements);
  return {
    payload: result.payload as ExactCasperPayload,
    payerAddress: clientSigner.accountAddress(),
  };
}

function buildPaymentPayload(payload: ExactCasperPayload): {
  x402Version: number;
  accepted: {
    scheme: "exact";
    network: string;
    asset: string;
    amount: string;
    payTo: string;
    maxTimeoutSeconds: number;
    extra: Record<string, unknown>;
  };
  payload: Record<string, unknown>;
} {
  return {
    x402Version: 2,
    accepted: {
      scheme: "exact",
      network: testNetwork,
      asset: testAsset,
      amount: "1000000",
      payTo: testPayTo,
      maxTimeoutSeconds: 300,
      extra: {
        name: "TestToken",
        version: "1",
      },
    },
    payload: payload as unknown as Record<string, unknown>,
  };
}

function buildRequirements(
  overrides?: Partial<{
    scheme: string;
    network: string;
    asset: string;
    amount: string;
    payTo: string;
    extra: Record<string, unknown>;
  }>,
): {
  scheme: string;
  network: string;
  asset: string;
  amount: string;
  payTo: string;
  maxTimeoutSeconds: number;
  extra: Record<string, unknown>;
} {
  return {
    scheme: "exact",
    network: testNetwork,
    asset: testAsset,
    amount: "1000000",
    payTo: testPayTo,
    maxTimeoutSeconds: 300,
    extra: {
      name: "TestToken",
      version: "1",
    },
    ...overrides,
  };
}

describe("ExactCasperScheme facilitator", () => {
  describe("getExtra", () => {
    it("returns feePayer and publicKey", async () => {
      const signer = createMockSigner();
      const scheme = new ExactCasperScheme(signer);

      const extra = scheme.getExtra(testNetwork);
      const addresses = signer.getAddresses(testNetwork);

      expect(extra).toEqual({ feePayer: addresses[0] });
    });

    it("returns empty string feePayer when signer has no addresses", async () => {
      const signer = createMockSigner();
      signer.getAddresses = () => [];
      const scheme = new ExactCasperScheme(signer);

      const extra = scheme.getExtra(testNetwork);

      expect(extra).toEqual({ feePayer: "" });
    });
  });

  describe("getSigners", () => {
    it("returns signer addresses", async () => {
      const signer = createMockSigner();
      const scheme = new ExactCasperScheme(signer);

      const addresses = await scheme.getSigners(testNetwork);

      expect(addresses).toEqual(signer.getAddresses(testNetwork));
    });
  });

  describe("verify", () => {
    it("returns valid for a correct payload", async () => {
      const { payload } = await createValidPayload();
      const scheme = new ExactCasperScheme(createMockSigner());

      const result = await scheme.verify(buildPaymentPayload(payload), buildRequirements());

      expect(result.isValid).toBe(true);
      expect(result.payer).toBe(payload.authorization.from);
    });

    it("rejects wrong scheme", async () => {
      const { payload } = await createValidPayload();
      const scheme = new ExactCasperScheme(createMockSigner());

      const result = await scheme.verify(
        buildPaymentPayload(payload),
        buildRequirements({ scheme: "upto" }),
      );

      expect(result.isValid).toBe(false);
      expect(result.invalidReason).toBe(ErrInvalidScheme);
    });

    it("rejects network mismatch", async () => {
      const { payload } = await createValidPayload();
      const scheme = new ExactCasperScheme(createMockSigner());

      const result = await scheme.verify(
        buildPaymentPayload(payload),
        buildRequirements({ network: "casper:other" }),
      );

      expect(result.isValid).toBe(false);
      expect(result.invalidReason).toBe(ErrNetworkMismatch);
    });

    it("rejects payTo mismatch", async () => {
      const { payload } = await createValidPayload();
      const scheme = new ExactCasperScheme(createMockSigner());

      const result = await scheme.verify(
        buildPaymentPayload(payload),
        buildRequirements({ payTo: "00".padEnd(66, "0") }),
      );

      expect(result.isValid).toBe(false);
      expect(result.invalidReason).toBe(ErrPayToMismatch);
    });

    it("rejects amount mismatch", async () => {
      const { payload } = await createValidPayload();
      const scheme = new ExactCasperScheme(createMockSigner());

      const result = await scheme.verify(
        buildPaymentPayload(payload),
        buildRequirements({ amount: "2000000" }),
      );

      expect(result.isValid).toBe(false);
      expect(result.invalidReason).toBe(ErrAmountMismatch);
    });

    it("rejects expired authorization", async () => {
      const { payload } = await createValidPayload();
      payload.authorization.validBefore = String(Math.floor(Date.now() / 1000) - 10);
      payload.authorization.validAfter = String(Math.floor(Date.now() / 1000) - 100);

      const scheme = new ExactCasperScheme(createMockSigner());
      const result = await scheme.verify(buildPaymentPayload(payload), buildRequirements());

      expect(result.isValid).toBe(false);
      expect(result.invalidReason).toBe(ErrExpired);
    });

    it("rejects not yet valid authorization", async () => {
      const { payload } = await createValidPayload();
      payload.authorization.validAfter = String(Math.floor(Date.now() / 1000) + 100);
      payload.authorization.validBefore = String(Math.floor(Date.now() / 1000) + 200);

      const scheme = new ExactCasperScheme(createMockSigner());
      const result = await scheme.verify(buildPaymentPayload(payload), buildRequirements());

      expect(result.isValid).toBe(false);
      expect(result.invalidReason).toBe(ErrNotYetValid);
    });

    it("rejects when public key does not match authorization.from", async () => {
      const { payload } = await createValidPayload();
      const otherKey = PrivateKey.generate(KeyAlgorithm.ED25519);
      payload.publicKey = otherKey.publicKey.toHex();

      const scheme = new ExactCasperScheme(createMockSigner());
      const result = await scheme.verify(buildPaymentPayload(payload), buildRequirements());

      expect(result.isValid).toBe(false);
      expect(result.invalidReason).toBe(ErrPublicKeyMismatch);
    });

    it("rejects invalid signature", async () => {
      const { payload } = await createValidPayload();
      payload.signature = "01" + "0".repeat(128);

      const scheme = new ExactCasperScheme(createMockSigner());
      const result = await scheme.verify(buildPaymentPayload(payload), buildRequirements());

      expect(result.isValid).toBe(false);
      expect(result.invalidReason).toBe(ErrInvalidSignature);
    });

    it("rejects invalid asset", async () => {
      const { payload } = await createValidPayload();
      const scheme = new ExactCasperScheme(createMockSigner());

      const result = await scheme.verify(
        buildPaymentPayload(payload),
        buildRequirements({ asset: "bad" }),
      );

      expect(result.isValid).toBe(false);
      expect(result.invalidReason).toBe(ErrInvalidAsset);
    });

    it("rejects invalid payTo", async () => {
      const { payload } = await createValidPayload();
      payload.authorization.to = "bad";
      const scheme = new ExactCasperScheme(createMockSigner());

      const result = await scheme.verify(
        buildPaymentPayload(payload),
        buildRequirements({ payTo: "bad" }),
      );

      expect(result.isValid).toBe(false);
      expect(result.invalidReason).toBe(ErrInvalidPayTo);
    });
  });

  describe("settle", () => {
    it("returns success for a valid payload", async () => {
      const { payload } = await createValidPayload();
      const signer = createMockSigner();
      const scheme = new ExactCasperScheme(signer);

      const result = await scheme.settle(buildPaymentPayload(payload), buildRequirements());

      expect(result.success).toBe(true);
      expect(result.transaction).toBe("transaction-hash");
      expect(result.network).toBe(testNetwork);
      expect(result.payer).toBe(payload.authorization.from);
    });

    it("fails when verify fails", async () => {
      const { payload } = await createValidPayload();
      const scheme = new ExactCasperScheme(createMockSigner());

      const result = await scheme.settle(
        buildPaymentPayload(payload),
        buildRequirements({ amount: "2000000" }),
      );

      expect(result.success).toBe(false);
      expect(result.errorReason).toBe(ErrAmountMismatch);
    });

    it("fails when putTransaction fails", async () => {
      const { payload } = await createValidPayload();
      const signer = createMockSigner({
        putTransaction: async () => {
          throw new Error("rpc error");
        },
      });
      const scheme = new ExactCasperScheme(signer);

      const result = await scheme.settle(buildPaymentPayload(payload), buildRequirements());

      expect(result.success).toBe(false);
      expect(result.errorReason).toBe(ErrSettleFailed);
      expect(result.errorMessage).toContain("rpc error");
    });

    it("fails when waitForTransaction fails", async () => {
      const { payload } = await createValidPayload();
      const signer = createMockSigner({
        waitForTransaction: async () => {
          throw new Error("timeout");
        },
      });
      const scheme = new ExactCasperScheme(signer);

      const result = await scheme.settle(buildPaymentPayload(payload), buildRequirements());

      expect(result.success).toBe(false);
      expect(result.errorReason).toBe(ErrSettleFailed);
      expect(result.errorMessage).toContain("timeout");
    });

    it("signs and submits a TransactionV1", async () => {
      const { payload } = await createValidPayload();
      const signed: TransactionV1[] = [];
      const submitted: TransactionV1[] = [];

      const signer = createMockSigner({
        signTransaction: async transaction => {
          signed.push(transaction);
        },
        putTransaction: async (_network, transaction) => {
          submitted.push(transaction);
          return "deploy-hash";
        },
      });

      const scheme = new ExactCasperScheme(signer);
      const result = await scheme.settle(buildPaymentPayload(payload), buildRequirements());

      expect(result.success).toBe(true);
      expect(signed.length).toBe(1);
      expect(submitted.length).toBe(1);
      expect(signed[0]).toBe(submitted[0]);
    });
  });
});

File: go/x402/mechanisms/casper/exact/facilitator/scheme_test.go
package facilitator_test

import (
	"context"
	"fmt"
	"strings"
	"testing"
	"time"

	"casper_x402_facilitator/x402/mechanisms/casper"
	casperFacilitator "casper_x402_facilitator/x402/mechanisms/casper/exact/facilitator"

	caspertypes "github.com/make-software/casper-go-sdk/v2/types"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
	x402 "github.com/x402-foundation/x402/go"
	"github.com/x402-foundation/x402/go/types"
)

const (
	testNetwork     = casper.NetworkCasperTestnet
	testAsset       = "aabbccddeeff0011223344556677889900aabbccddeeff001122334455667788"
	testPayTo       = "00aabbccddeeff0011223344556677889900aabbccddeeff001122334455667788"
	testFrom        = "01aabbccddeeff0011223344556677889900aabbccddeeff001122334455667788"
	facilitatorAddr = testFrom
)

type mockFacilitatorSigner struct {
	addresses       []string
	publicKeyHex    string
	verifyResult    bool
	verifyErr       error
	putDeployResult string
	putDeployErr    error
	waitErr         error
}

func (m *mockFacilitatorSigner) GetNetworkConfig(_ context.Context, _ string) (casper.NetworkConfig, error) {
	return casper.NetworkConfig{
		testNetwork,
		"http://127.0.0.1:11101/rpc",
	}, nil
}

func (m *mockFacilitatorSigner) GetAddresses(_ context.Context, _ string) []string {
	return m.addresses
}

func (m *mockFacilitatorSigner) GetPublicKeyHex(_ context.Context, _ string) (string, error) {
	return m.publicKeyHex, nil
}

func (m *mockFacilitatorSigner) VerifyEIP712Signature(_ [32]byte, _ [65]byte, _ string) (bool, error) {
	return m.verifyResult, m.verifyErr
}

func (m *mockFacilitatorSigner) SignTransaction(_ *caspertypes.TransactionV1, _ string) error {
	return nil
}
func (m *mockFacilitatorSigner) PutTransaction(_ context.Context, _ string, _ caspertypes.TransactionV1) (string, error) {
	return m.putDeployResult, m.putDeployErr
}
func (m *mockFacilitatorSigner) WaitForTransaction(_ context.Context, _ string, _ string) error {
	return m.waitErr
}

func defaultSigner() *mockFacilitatorSigner {
	return &mockFacilitatorSigner{
		addresses:       []string{facilitatorAddr},
		publicKeyHex:    "01aabbccddeeff0011223344556677889900aabbccddeeff001122334455667788",
		verifyResult:    true,
		putDeployResult: strings.Repeat("ab", 32),
	}
}

func buildValidPayload(t *testing.T) (types.PaymentPayload, types.PaymentRequirements) {
	t.Helper()

	now := time.Now().Unix()
	payloadObj := &casper.ExactCasperPayload{
		Signature: strings.Repeat("ab", 65),
		PublicKey: "01aabbccddeeff0011223344556677889900aabbccddeeff001122334455667788",
		Authorization: casper.ExactCasperAuthorization{
			From:        testFrom,
			To:          testPayTo,
			Value:       "1000000",
			ValidAfter:  fmt.Sprintf("%d", now-60),
			ValidBefore: fmt.Sprintf("%d", now+300),
			Nonce:       "aabbccddeeff0011223344556677889900aabbccddeeff001122334455667788",
		},
	}

	payload := types.PaymentPayload{
		X402Version: 2,
		Payload:     payloadObj.ToMap(),
		Accepted: types.PaymentRequirements{
			Scheme:  "exact",
			Network: testNetwork,
		},
	}

	req := types.PaymentRequirements{
		Scheme:  "exact",
		Network: testNetwork,
		Asset:   testAsset,
		Amount:  "1000000",
		PayTo:   testPayTo,
		Extra: map[string]interface{}{
			"name":    "TestToken",
			"version": "1",
		},
	}

	return payload, req
}

func TestGetExtra_ReturnsCasperMetadata(t *testing.T) {
	signer := defaultSigner()
	scheme := casperFacilitator.NewExactCasperScheme(signer, nil)

	extra := scheme.GetExtra(x402.Network(testNetwork))
	require.NotNil(t, extra)
	assert.Equal(t, facilitatorAddr, extra["feePayer"])
}

func TestGetExtra_EmptySignerAddresses(t *testing.T) {
	signer := defaultSigner()
	signer.addresses = nil
	scheme := casperFacilitator.NewExactCasperScheme(signer, nil)

	extra := scheme.GetExtra(x402.Network(testNetwork))
	require.NotNil(t, extra)
	assert.Equal(t, "", extra["feePayer"])
}

func TestVerify_HappyPath(t *testing.T) {
	signer := defaultSigner()
	scheme := casperFacilitator.NewExactCasperScheme(signer, nil)

	payload, req := buildValidPayload(t)
	resp, err := scheme.Verify(context.Background(), payload, req, nil)
	require.NoError(t, err)
	assert.True(t, resp.IsValid)
	assert.Equal(t, testFrom, resp.Payer)
}

func TestVerify_WrongScheme(t *testing.T) {
	signer := defaultSigner()
	scheme := casperFacilitator.NewExactCasperScheme(signer, nil)

	payload, req := buildValidPayload(t)
	payload.Accepted.Scheme = "upto"

	_, err := scheme.Verify(context.Background(), payload, req, nil)
	assert.ErrorContains(t, err, casperFacilitator.ErrUnsupportedScheme)
}

func TestVerify_NetworkMismatch(t *testing.T) {
	signer := defaultSigner()
	scheme := casperFacilitator.NewExactCasperScheme(signer, nil)

	payload, req := buildValidPayload(t)
	payload.Accepted.Network = casper.NetworkCasperMainnet

	_, err := scheme.Verify(context.Background(), payload, req, nil)
	assert.ErrorContains(t, err, casperFacilitator.ErrNetworkMismatch)
}

func TestVerify_PayToMismatch(t *testing.T) {
	signer := defaultSigner()
	scheme := casperFacilitator.NewExactCasperScheme(signer, nil)

	payload, req := buildValidPayload(t)
	req.PayTo = "00" + strings.Repeat("bb", 32)

	_, err := scheme.Verify(context.Background(), payload, req, nil)
	assert.ErrorContains(t, err, casperFacilitator.ErrPayToMismatch)
}

func TestVerify_AmountMismatch(t *testing.T) {
	signer := defaultSigner()
	scheme := casperFacilitator.NewExactCasperScheme(signer, nil)

	payload, req := buildValidPayload(t)
	req.Amount = "9999999"

	_, err := scheme.Verify(context.Background(), payload, req, nil)
	assert.ErrorContains(t, err, casperFacilitator.ErrAmountMismatch)
}

func TestVerify_Expired(t *testing.T) {
	signer := defaultSigner()
	scheme := casperFacilitator.NewExactCasperScheme(signer, nil)

	payload, req := buildValidPayload(t)
	auth := payload.Payload["authorization"].(map[string]interface{})
	auth["validBefore"] = fmt.Sprintf("%d", time.Now().Unix()-1)

	_, err := scheme.Verify(context.Background(), payload, req, nil)
	assert.ErrorContains(t, err, casperFacilitator.ErrPayloadExpired)
}

func TestVerify_NotYetValid(t *testing.T) {
	signer := defaultSigner()
	scheme := casperFacilitator.NewExactCasperScheme(signer, nil)

	payload, req := buildValidPayload(t)
	now := time.Now().Unix()
	auth := payload.Payload["authorization"].(map[string]interface{})
	auth["validAfter"] = fmt.Sprintf("%d", now+9999)
	auth["validBefore"] = fmt.Sprintf("%d", now+10000)

	_, err := scheme.Verify(context.Background(), payload, req, nil)
	assert.ErrorContains(t, err, casperFacilitator.ErrNotYetValid)
}

func TestSettle_HappyPath(t *testing.T) {
	signer := defaultSigner()
	scheme := casperFacilitator.NewExactCasperScheme(signer, nil)

	payload, req := buildValidPayload(t)
	resp, err := scheme.Settle(context.Background(), payload, req, nil)
	require.NoError(t, err)
	assert.True(t, resp.Success)
	assert.Regexp(t, `^[0-9a-fA-F]{64}$`, resp.Transaction)
	assert.Equal(t, testFrom, resp.Payer)
	assert.Equal(t, x402.Network(testNetwork), resp.Network)
}

func TestSettle_VerifyFails(t *testing.T) {
	signer := defaultSigner()
	signer.verifyResult = false

	scheme := casperFacilitator.NewExactCasperScheme(signer, nil)
	payload, req := buildValidPayload(t)

	_, err := scheme.Settle(context.Background(), payload, req, nil)
	assert.ErrorContains(t, err, casperFacilitator.ErrVerificationFailed)
}

func TestSettle_PutDeployFails(t *testing.T) {
	signer := defaultSigner()
	signer.putDeployErr = fmt.Errorf("rpc timeout")

	scheme := casperFacilitator.NewExactCasperScheme(signer, nil)
	payload, req := buildValidPayload(t)

	_, err := scheme.Settle(context.Background(), payload, req, nil)
	assert.ErrorContains(t, err, casperFacilitator.ErrPutDeployFailed)
}

func TestSettle_WaitDeployFails(t *testing.T) {
	signer := defaultSigner()
	signer.waitErr = fmt.Errorf("context deadline exceeded")

	scheme := casperFacilitator.NewExactCasperScheme(signer, nil)
	payload, req := buildValidPayload(t)

	_, err := scheme.Settle(context.Background(), payload, req, nil)
	assert.ErrorContains(t, err, casperFacilitator.ErrWaitDeployFailed)
}
